WIP: Initialize keymap library.

Ambrevar opened the pull request:
Do not merge.

Important changes:

- Moved to a separate library.
- Multiple inheritance.
- `define-key` validates the keyspec at compile time (to be tested... done).
- Prefix keymaps.
- Shorter, simpler code!

Left to do:

- [x] Tests!
- [x] Implement keycode keyspec syntax (e.g. "#10").
- [x] Implement default key translation function.
- [x] Detect cycles in lookup-key.
- [x] List keys?
- [x] List all bindings associated to a symbol.
- [x] Exports.
- [x] Fallback value when not found.
- [x] Integrate.
- [x] Fix dead keys if they still don't work in the minibuffer.
- [x] Test with SBCL 2.0.2 (has stronger warnings)
- [ ] Move `scheme` support to `keymap` library.
- [x] Use new string protocol (all lower case)
- [ ] Implement CUA bindings, make schemes inherit from each other.
- [x] Test key code support (try switching system keyboard layout)
- [x] Document how modifiers can be remapped.

Ambrevar commented:
@jmercouris This is ready for review!

In a nutshell:
- 200 lines of code remove from source!
- The input system is super simple now (compared to the convoluted stuff we inherited from the RPC architecture).
- SBCL 2.0.2 compatible.
- Lots of new features, including keymap inheritance, prefix keymap, composition, etc.
- Library should be well tested, documented, typed, etc.  Robust and hopefully future-proof.

jmercouris commented:
Yay!!! I will look at this as soon as I can tomorrow!

Ambrevar commented:
Thanks!

A few TODO to consider:

- Can a keycode be 0?  I've used 0 as the unset value.

- How shall we call the strings like "C-c C-d"?  I've called them
  `keyspec`s, but what about `keydesc` instead?

- The function `lookup-key` is named after its Emacs counterpart, but
  really it should be `lookup-keys` for consistency with the rest of the
  code.  Thoughts?

- symbol-keys returns an alist of (KEYSPEC KEYMAP), which might be more
  usable than a hash-table since it chains well with #'mapcar, etc.
  Do we really need the KEYMAPs though?  In Next we don't.
  Shall we return them as second VALUE?

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+  (shortcut "" :type string))
+
+(declaim (ftype (function ((or string modifier) (or string modifier)) boolean) modifier=))
+(defun modifier= (string-or-modifier1 string-or-modifier2)
```
Is there a possibility to just always convert to a string for comparison?

jmercouris replied:
I am just thinking it may simplify the logic, but there are still many things I could be missing.

jmercouris replied:
Or would it be possible to use `modspec->modifier`?

Ambrevar replied:
Can't use modspec->modifier because it relies on `modifier=`.

Converting to string would result in the same logic I think.  If you can think of a simplification, please share it :)

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+                    (modifier-shortcut string-or-modifier2)))))))
+
+(defmethod fset:compare ((x modifier) (y modifier))
+  "Needed to user the KEY structure as keys in Fset maps."
```
Probable typo on "user"

Ambrevar replied:
Fixed.

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+(defvar +super+ (make-modifier :string "super" :shortcut "S"))
+(defvar +hyper+ (make-modifier :string "hyper" :shortcut "H"))
+
+(defparameter *modifier-list*
```
Probably also want to make this +modifier-list+ since it is not meant to be mutated?

jmercouris replied:
Or, is the idea that people could append any modifiers that they want to this list?

Ambrevar replied:
That latter indeed!

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+(deftype key-status-type ()
+  `(or (eql :pressed) (eql :released)))
+
+;; Must be a structure so that it can served as a key in a hash-table with
```
Is the problem of the case insensitivity of equalp solved elsewhere?

Ambrevar replied:
See `fset:compare key key`.

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+  (status :pressed :type key-status-type))
+
+(defun key= (key1 key2)
+  "Two keys are equal if the have the same modifiers, status and key code.
```
Is the value the printed representation of the character? I assume based on the struct, but I am not fully sure :-D

Ambrevar replied:
Yes, couldn't think of a better name... :(

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+  "Return new `key'.
+Modifiers can be either a `modifier' type or a string that will be looked up in
+`*modifier-list*'."
+  ;; TODO: Display warning on duplicate modifiers?
```
it would be nice I guess, but how far do we go? There are an infinite amount of error permutations

Ambrevar replied:
I think you misunderstood what I mean: we just need to report extraneous modifiers, e.g. in ("C" "C").
Anyways, fixed in e195556070102165a935479b6c93346d93147082.

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+;; #'equalp tests.
+(defstruct (key (:constructor %make-key (code value modifiers status))
+                (:copier %copy-key))
+  (code 0 :type integer) ; TODO: Can a keycode be 0?  I think not, so 0 might be a good non-value.
```
I was not able to find anything with a value of 0 other than Nul, which I don't think one can enter

Ambrevar replied:
Thanks.

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+                               (:modifiers fset:wb-set) (:status keyword))
+                          key)
+                copy-key))
+(defun copy-key (key &key (code (key-code key)) (value (key-value key))
```
I assume this exists because %copy-key does not recurse (as far as I understand based on the Hyperspec)?

Ambrevar replied:
Not exactly: the copier is automatically defined by `defstruct`, so if I want to override it, I need to give vanilla copier another name and implement my own `copy-key` that uses it.

jmercouris replied:
Right, but my question is, why did you want to override it?

Ambrevar replied:
To support modspecs argument.

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+(declaim (ftype (function (string) list-of-keys) keyspecs->keys))
+(defun keyspecs->keys (spec)
+  "Parse SPEC and return corresponding list of keys."
+  ;; TODO: Return nil if SPEC is invalid?
```
A condition will be raised upon each invalid key, should the user pick a restart that allows them to return a valid list of keys, then it should be OK, no?

jmercouris replied:
s/invalid key/invalid keyspec->key invokation

Ambrevar replied:
I'm not sure I understood what you meant: that we added a restart?
I'm not sure a restart is too useful here though.  This function is internal to `keymap` and it's only called by `define-key`, which is a macro and thus detects errors at compile time.  Restarts are not very useful during compilation.

jmercouris replied:
We don't have to add a restart, the user could make some changes and use one of the default restarts, I could for example edit the define-key form and attempt a recompile

Ambrevar replied:
Then the current code behaves well, doesn't it?

jmercouris replied:
Yes, that is what I was saying, I think it just fine!

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+                (copy-key key :modifiers (fset:less (key-modifiers key) +shift+)))
+              keys))))
+
+(defun translate-remove-but-first-control (keys)
```
What if we want to bind C-x C-c and C-x c to different things? I guess it doesn't matter since it doesn't appear in the default translator, maybe some people will want it :-)

Ambrevar replied:
It's supported: translation only happens if the exact keyspec does not match.

jmercouris commented on source/renderer-gtk.lisp:
```diff
@@ -86,12 +86,8 @@
-     gtk-object "key_release_event"
-     (lambda (widget event) (declare (ignore widget))
-       (pop-modifier *browser* event)))
+    ;; REVIEW: No need to handle release events?
```
Correct, no need to handle release events. Insofar as I can tell most UI elements (exception being buttons), the click on matters, not the click off.

jmercouris commented on source/renderer-gtk.lisp:
```diff
@@ -148,63 +153,39 @@
-         (key-string (format nil "button~s" button))
-         (window (find sender (window-list)
-                       :key #'active-buffer)))
+         ;; REVIEW: No need to store X and Y?
```
I do think we should remove it since we are not using it and have not yet found cause to use it, yes

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+(defvar *default-translator* #'translate-shift-control-combinations
+  "Default key translator to use in `keymap' objects.")
+
+;; TODO: Enable override of default and translator in lookup?
```
Yes, I can think of when it would be useful in many cases! :-)

jmercouris replied:
That said, someone could just use `let`... but maybe if it is sanctioned with an &key people will feel more comfortable doing it

Ambrevar replied:
`let` would not work here since `default` and `translator` are `keymap` slots.
I'll use a keyword argument.

jmercouris replied:
Ah, whoops :-)

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+(defun keymap-p (object)
+  (typep object 'keymap))
+
+(deftype keyspecs-type ()               ; TODO: Rename to KEYDESC?
```
In my opinion it is a specification not a description. I could describe the keys in a number of ways, but I can only provide an official representation via a specification. The connotation being that a specification is a logically sound description, a subset of a description.

Ambrevar replied:
Fair enough, thanks for making a point here! :)

jmercouris commented on libraries/keymap/keymap.lisp:
```diff
@@ -0,0 +1,556 @@
+  `(satisfies keyspecs->keys))
+
+;; We need a macro to check that bindings are valid at compile time.
+;; This is because most Common Lisp implementations or not capable of checking
```
s/or not/are not

jmercouris commented:
Well, well! Pierre! I must say a most fantastic job! The functions are clear, simple and concise! Easy to follow logic! Even though I did not have you walk me through the codebase, I still understand a large part of it. There are some things I must learn like `fset` before I can leave a full review.

Though I'm not yet fully understanding this part of the codebase, I left some preliminary comments to start a discussion.

Lastly (for now), there is a form that does not compile: 
```
../../source/vi-mode.lisp:37:7:
  error: 
    during macroexpansion of (DEFINE-KEY :KEYMAP MAP ...). Use *BREAK-ON-SIGNALS*
    to intercept.
    
     The value
       MAP
     is not of type
       STRING
     when binding KEYMAP::SPEC
```

jmercouris commented:
Some quick thoughts:
>- The function `lookup-key` is named after its Emacs counterpart, but really it should be `lookup-keys` for consistency with the rest of the code. Thoughts?

Let's go with `lookup-keys`, we do not have to be 1:1 Emacs, we already differ in so many concepts

> - symbol-keys returns an alist of (KEYSPEC KEYMAP), which might be more usable than a hash-table since it chains well with #'mapcar, etc. Do we really need the KEYMAPs though? In Next we don't. Shall we return them as second VALUE?

A second VALUE should be fine in this case. Again though, don't forget about `maphash` :-)

Ambrevar commented:
vi-mode: Oops!  Fixed.  Weird I didn't see it though :p 

lookup-keys: OK.

maphash: The big drawback of maphash is that it does not return a value.  In general it's nice to return values that can be passed to higher-order functions.
I think I'll go with a (VALUES LIST ALIST) return value.

Ambrevar commented:
Actually about lookup-key: we have define-key without `s`, so lookup-key is consistent with that at least.

Ambrevar commented:
Fset tutorial: https://common-lisp.net/project/fset/Site/FSet-Tutorial.html

jmercouris commented:
By the way, feel free to merge this whenever you feel ready, in so far as I am concerned, the design is fully sound! The only things I could make comments would be trivial stylistic choices.

Ambrevar commented:
I've made some structural changes, so I'd like you to let me know what you think
before I merge:

- Made keymap-local translators  global: indeed, keymap-specific translators
  would not do the right thing (e.g. "g g" and "G" are bound to different
  keymaps, "g g" is in the first keymap, "G" in the second, if the user presses
  "s-G" it used to hit "g" instead of "G").
  
- Removed DEFAULT slot since we are not using it, plus since LOOKUP-KEY supports
  multiple keymaps as argument, there is no intuitive way to use DEFAULT.

- Added a mandatory NAME slot for keymaps.  This is useful to describe the
  keymap.
  Try "command-inspect" on a command that has bindings!

Also great news, dead keys are finally supported in the minibuffer.

jmercouris commented on documents/CHANGELOG.org:
```diff
@@ -39,6 +39,15 @@ MAJOR.MINOR.PATCH, we increment the:
 
+** 2.0.0
+- Change key string specifications: all lowercase without underscore.
+  For instance "Page_Up" becomes "pageup".
```
I think it should be case insensitive actually if the user has C-SPACE or C-space it should not matter. Thoughts?

Ambrevar commented:
The problem is that key "A" is different from "a".  So if we want to be
case insensitive for only some keys, we need a way to distinguish them.

I guess the criteria here is "printability".  Although it does not work
for "space", "hyphen", "tab" and maybe a few others which are printable.

The drawbacks may outweigh the benefits.

jmercouris commented on source/start.lisp:
```diff
@@ -201,11 +201,8 @@ EXPR must contain one single Lisp form. Use `progn' if needed."
-                                                   (open-urls urls)
-                                                   ;; Return nil so that the callback is called only once.
-                                                   nil)))
+                       (gtk:within-gtk-thread
```
I am pretty sure this will cause freezing on macOS

jmercouris replied:
Actually I can't even find the symbol within-gtk-thread in cl-cffi-gtk

jmercouris replied:
I changed it to within-main-loop, doesn't appear to be causing problems

Ambrevar replied:
It's in GTK init.
It will work on macOS, because the macro does exactly what the previous code did:
```
(defmacro within-gtk-thread (&body body)
  "Executes the body within the GTK thread. Unlike GTK:WITHIN-MAIN-LOOP, this
macro does not initialize GTK, and can be used several times in a program. It
is mainly useful to execute some GTK code from a thread different from the main
GTK thread."
  `(gdk:gdk-threads-add-idle ; Could probably be replaced by glib:g-idle-add, but this is safer (see GDK Threads doc)
    #'(lambda ()
        ,@body
        nil)))
```

Ambrevar replied:
I meant `gtk-init.lisp` from the CFFI-GTK library.

Ambrevar replied:
You _cannot_ use within-main-loop here, look at the documentation of `within-gtk-thread`:
```
  "Executes the body within the GTK thread. Unlike GTK:WITHIN-MAIN-LOOP, this
macro does not initialize GTK, and can be used several times in a program. It
is mainly useful to execute some GTK code from a thread different from the main
GTK thread."
```

jmercouris replied:
Does not exist in my version of cl-cffi-gtk

jmercouris replied:
I had to update to the one on Quicklisp, I was using the one from crategus cloned locally

jmercouris commented:
> The problem is that key "A" is different from "a". So if we want to be case insensitive for only some keys, we need a way to distinguish them. I guess the criteria here is "printability". Although it does not work for "space", "hyphen", "tab" and maybe a few others which are printable. The drawbacks may outweigh the benefits.

I know you are pro lower case, but I am pro upper case, it would seem to me that the best compromise would be using the lisp printer settings :-D

jmercouris commented:
> I've made some structural changes, so I'd like you to let me know what you think
> before I merge:
> 
> * Made keymap-local translators  global: indeed, keymap-specific translators
>   would not do the right thing (e.g. "g g" and "G" are bound to different
>   keymaps, "g g" is in the first keymap, "G" in the second, if the user presses
>   "s-G" it used to hit "g" instead of "G").

Ah, I did not even think of this potential conflict! In any case it will now allow for let binding it no?

> * Removed DEFAULT slot since we are not using it, plus since LOOKUP-KEY supports
>   multiple keymaps as argument, there is no intuitive way to use DEFAULT.

Agreed

> * Added a mandatory NAME slot for keymaps.  This is useful to describe the
>   keymap.

This is a good change!

> Also great news, dead keys are finally supported in the minibuffer.

Yay!

jmercouris commented:
Just tested on macOS, I cannot type M-x, M->, etc, so I cannot test describe-* :-\

Ambrevar commented:
Can you report precisely what does not work?  If you enable debug log, you should see exactly what's happening.

Also check `translate-modifiers`, I've updated the function a bit, maybe I broke something with the magic macOS modifier number.

Ambrevar commented:
> I know you are pro lower case, but I am pro upper case, it would seem to me that the best compromise would be using the lisp printer settings :-D

Unfortunately these are strings, not symbols, so the Lisp printer does
not intervene here.

Ambrevar commented:
> Ah, I did not even think of this potential conflict! In any case it will now allow for let binding it no?

Correct!

Ambrevar commented:
> Just tested on macOS, I cannot type M-x, M->, etc, so I cannot test describe-* :-\

If you have a init.lisp, make sure it's compatible with the changes :p

Ambrevar commented:
I've pushed some last tweaks.  This is ready to merge as far as I'm concerned.
Let me know if it works for you and when it does, let's merge!

jmercouris commented:
nothing is in my init file.

jmercouris commented:
Here's what it does when I type in M-x:

![Screen Shot 2020-03-20 at 18 09 15](https://user-images.githubusercontent.com/1691662/77188321-18cf3600-6ad6-11ea-9746-3acb44e043f2.png)

Also right clicking crashes the program:

```
debugger invoked on a TYPE-ERROR in thread
#<THREAD "main thread" RUNNING {1000560083}>:
  The value
    0
  is not of type
    LIST
  when binding NEXT::MODIFIER-STATE

Type HELP for debugger help, or (SB-EXT:EXIT) to exit from SBCL.

restarts (invokable by number or by possibly-abbreviated name):
  0: [RETURN-FROM-G-CLOSURE] Return value from closure
  1: [ABORT                ] Exit debugger, returning to top level.

(NEXT::TRANSLATE-MODIFIERS 0 #S(GDK:GDK-EVENT-BUTTON :TYPE :BUTTON-PRESS :WINDOW #<GDK:GDK-WINDOW {1005E3E383}> :SEND-EVENT NIL :TIME 4176280826 :X 273.75d0 :Y 459.62109375d0 :AXES #(273.75d0 459.62109375d0) :STATE 0 :BUTTON 3 :DEVICE #<GDK:GDK-X11-DEVICE-XI2 {1005F3D883}> :X-ROOT 273.75d0 :Y-ROOT 481.62109375d0)) [external]
   source: (DEFUN TRANSLATE-MODIFIERS (MODIFIER-STATE &OPTIONAL EVENT)
             "Return list of modifiers fit for `keymap:make-key'.
See `gtk-browser's `modifier-translator' slot."
             (LET ((MODIFIERS NIL)
                   (MACINTOSH-ALT-KEY-VALUE 65406)
                   (KEY-VALUE
                    (MATCH (TYPE-OF EVENT) ('GDK:GDK-EVENT-KEY #) (_ 0))))
               (WHEN (MEMBER :CONTROL-MASK MODIFIER-STATE)
                 (PUSH "control" MODIFIERS))
               (WHEN
                   (OR (MEMBER :MOD1-MASK MODIFIER-STATE)
                       (= KEY-VALUE MACINTOSH-ALT-KEY-VALUE))
                 (PUSH "meta" MODIFIERS))
               (WHEN (MEMBER :SHIFT-MASK MODIFIER-STATE)
                 (PUSH "shift" MODIFIERS))
               (WHEN
                   (OR (MEMBER :SUPER-MASK MODIFIER-STATE)
                       (AND (MEMBER :MOD2-MASK MODIFIER-STATE)
                            (MEMBER :META-MASK MODIFIER-STATE)))
                 (PUSH "super" MODIFIERS))
               (WHEN (MEMBER :HYPER-MASK MODIFIER-STATE)
                 (PUSH "hyper" MODIFIERS))
               MODIFIERS))
0] 
```

jmercouris commented:
Here's the debug output:
```
NEXT> (log:config :debug)
; No value<DEBUG> [18:14:54] next renderer-gtk.lisp (process-key-press-event gtk-window) -
  NEXT::KEY-STRING: "arabicswitch" NEXT::KEYCODE: 66 CHARACTER: #\Nul
  NEXT::KEYVAL-NAME: "Arabic_switch" NEXT:MODIFIERS: ("meta") 
<DEBUG> [18:14:54] next input.lisp (dispatch-input-event) -
  Forward key "M-arabicswitch (M-#66)"
<DEBUG> [18:14:54] next renderer-gtk.lisp (process-key-press-event gtk-window) -
  NEXT::KEY-STRING: "x" NEXT::KEYCODE: 15 CHARACTER: #\x NEXT::KEYVAL-NAME: "x" 
<DEBUG> [18:14:54] next input.lisp (dispatch-input-event) -
  Forward key "x (#15)"
<DEBUG> [18:14:55] next renderer-gtk.lisp (process-key-press-event gtk-window) -
  NEXT::KEY-STRING: "arabicswitch" NEXT::KEYCODE: 66 CHARACTER: #\Nul
  NEXT::KEYVAL-NAME: "Arabic_switch" NEXT:MODIFIERS: ("meta") 
<DEBUG> [18:14:55] next input.lisp (dispatch-input-event) -
  Forward key "M-arabicswitch (M-#66)"
```

Ambrevar commented:
Clean the .fasls (CFFI and Next), fetch the latest commit from this
branch and rebuild.

> NEXT> (log:config :debug)
> ; No value<DEBUG> [18:14:54] next renderer-gtk.lisp (process-key-press-event gtk-window) -
>   NEXT::KEY-STRING: "arabicswitch" NEXT::KEYCODE: 66 CHARACTER: #\Nul
>   NEXT::KEYVAL-NAME: "Arabic_switch" NEXT:MODIFIERS: ("meta") 

Which key did you press?

jmercouris commented:
I pressed the "alt" key, as in "alt-x" or "M-x"

Ambrevar commented:
> Here's what it does when I type in M-x:

Are you saying that typing "M-x" when the browser starts brings up the
"command-inspect" view of execute-command?

jmercouris commented:
> Here's what it does when I type in M-x:
> Are you saying that typing "M-x" when the browser starts brings up the "command-inspect" view of execute-command?

No, I typed in `C-h k`, then I typed in `M-x` so that you could see how it reports Meta/Alt on my machine.

jmercouris commented:
I cleaned all fasls, recompiled, issue persists unfortunately :-(

Ambrevar commented:
> No, I typed in `C-h k`, then I typed in `M-x` so that you could see how it reports Meta/Alt on my machine.
>

Then key-inspect works for you! :)

Ambrevar commented:
> I pressed the "alt" key, as in "alt-x" or "M-x"

I've just added Arabic_switch in the last commit, try again.

For crash with click and left/right, please report for each input key:

- the key,
- the debug output or the trace.

Ambrevar commented:
Which issue?

jmercouris commented on source/utility.lisp:
```diff
@@ -96,60 +78,11 @@ The second value is the initfunction."
-      (log:error "Process with PID ~a is not running" pid))))
-
-(declaim (ftype (function (string &rest string)) run-program-to-string))
-(defun run-program-to-string (program &rest args)
```
Is this not used anywhere?

Ambrevar replied:
Nope :p

jmercouris commented:
Still no luck:

output:

```
<DEBUG> [19:34:38] next renderer-gtk.lisp (process-key-press-event gtk-window) -
  NEXT::KEY-STRING: NIL NEXT::KEYCODE: 66 CHARACTER: #\Nul
  NEXT::KEYVAL-NAME: "Arabic_switch" NEXT:MODIFIERS: ("meta") 
<DEBUG> [19:34:38] next renderer-gtk.lisp (process-key-press-event gtk-window) -
  NEXT::KEY-STRING: "x" NEXT::KEYCODE: 15 CHARACTER: #\x NEXT::KEYVAL-NAME: "x" 
<DEBUG> [19:34:38] next input.lisp (dispatch-input-event) -
  Forward key "x (#15)"
<DEBUG> [19:34:38] next renderer-gtk.lisp (process-key-press-event gtk-window) -
  NEXT::KEY-STRING: "x" NEXT::KEYCODE: 15 CHARACTER: #\x NEXT::KEYVAL-NAME: "x" 
<DEBUG> [19:34:38] next input.lisp (dispatch-input-event) -
  Forward key "x (#15)"
```

now it just acts as if I am not pressing anything

jmercouris commented:
In other words, I do `C-h k`, and then I type in `M-x`, but it just prints `x` with no modifiers in the echo area

Ambrevar commented:
Looks like Meta (Alt) is not a modifier on macOS.
Is this why you implemented the push-modifiers / pop-modifiers functions before?
Can you check if 87b57377aedde7e161c30601c31bd6549fee0391
is the first commit where meta stops working for you?

Regarding your click issue, I think I've fixed it in the last commit (please pull).

jmercouris commented:
> Looks like Meta (Alt) is not a modifier on macOS.
> Is this why you implemented the push-modifiers / pop-modifiers functions before?
> Can you check if [87b5737](https://github.com/atlas-engineer/next/commit/87b57377aedde7e161c30601c31bd6549fee0391)
> is the first commit where meta stops working for you?
> 
> Regarding your click issue, I think I've fixed it in the last commit (please pull).

that's correct, thats why I implemented a function to push/pop modifiers :-)

it is a bug in GTK. You cannot query which modifiers exist (reliably), I thought I had documented that, sorry if I forgot!

Ambrevar commented:
Can you tell me which modifiers needs this form of stack?
Control seems to be fine.   How is Super?

jmercouris commented:
Right click and left click are now working! :-)

jmercouris commented:
I believe it is Meta, and Super. Only Control appears to behave reliably

jmercouris commented:
> Looks like Meta (Alt) is not a modifier on macOS.
> Is this why you implemented the push-modifiers / pop-modifiers functions before?
> Can you check if [87b5737](https://github.com/atlas-engineer/next/commit/87b57377aedde7e161c30601c31bd6549fee0391)
> is the first commit where meta stops working for you?

I can confirm that this is the first commit in which meta does not work.

jmercouris commented:
> I know you are pro lower case, but I am pro upper case, it would seem to me that the best compromise would be using the lisp printer settings :-D
> Unfortunately these are strings, not symbols, so the Lisp printer does not intervene here.

I'm know, I'm saying we let the lisp printer settings decide how we evaluate the strings

Ambrevar commented:
> I'm know, I'm saying we let the lisp printer settings decide how we evaluate the strings

In any case this is an easy fix, we can always do this later.

Ambrevar commented:
I've just pushed a fix for the modifiers on macOS.  Let me know if it works for you.

Ambrevar commented:
https://github.com/atlas-engineer/next/pull/599/commits/1eb5a1c9342465c20aa912cb4eb7ca974c3982a9

jmercouris commented:
Yay! It has fixed it!!!

I should also point out that bindings for:

```
"C-[" #'switch-buffer-previous
"C-]" #'switch-buffer-next
```

are no longer functioning because it now reports that as C-bracketleft and C-bracketright, not sure if that is intended or not. If it is intended, then we just need to change those bindings, and it should be ready for merging!

Ambrevar commented:
Actually all "C-" + punctuation bindings were broken.  Should be fixed now.

I've also cleaned up the macOS workaround, can you test again?   It should allow you to remap the modifiers very easily (example in the doc).

jmercouris commented:
Everything working! Yay :-)

Ambrevar commented:
Thanks!

jmercouris commented:
More bugs :-D, C-space does not work, does not actually do anything

Ambrevar commented:
C-space fixed.

Question: the keymap library supports "-" as a key value.
E.g. "-" and "C--" are valid.

Shall we still use "hyphen" or shall we use "-"?  The latter would
remove one exception and be more consistent with Emacs.

jmercouris commented:
C-- looks so much like a typo that I think C-hyphen is better, but either is OK with me

Ambrevar commented:
OK, C-hyphen it is then!
Big merge inbound! :)

jmercouris commented:
Let's go!!! :-)

jmercouris commented:
Can we close this pull request since it appears merged on master?

Ambrevar commented:
Yup!
